libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);
+			break;
+		}
+	}
+}
+
+bool ReferenceCollector::tryAddReference(Declaration const* _declaration, SourceLocation const& _location)
+{
+	if (&m_declaration != _declaration)
+		return false;
+
+	m_resultingReferences.emplace_back(_location, m_kind);
+	return true;
+}
+
+void ReferenceCollector::endVisit(Identifier const& _identifier)
+{
+	if (auto const* declaration = _identifier.annotation().referencedDeclaration)
+		tryAddReference(declaration, _identifier.location());
+
+	for (auto const* declaration: _identifier.annotation().candidateDeclarations + _identifier.annotation().overloadedDeclarations)
+		tryAddReference(declaration, _identifier.location());
+}
+
+
+void ReferenceCollector::endVisit(IdentifierPath const& _identifierPath)
+{
+	tryAddReference(_identifierPath.annotation().referencedDeclaration, _identifierPath.location());
+}
+
+void ReferenceCollector::endVisit(MemberAccess const& _memberAccess)
+{
+	if (_memberAccess.annotation().referencedDeclaration == &m_declaration)
+		m_resultingReferences.emplace_back(_memberAccess.location(), m_kind);
+}
+
+bool ReferenceCollector::visit(Assignment const& _node)
+{
+	auto const restoreKind = ScopeGuard{[this, savedKind=m_kind]() { m_kind = savedKind; }};
+
+	m_kind = DocumentHighlightKind::Write;
+	_node.leftHandSide().accept(*this);
+
+	m_kind = DocumentHighlightKind::Read;
+	_node.rightHandSide().accept(*this);
+
+	return false;
+}
+
+bool ReferenceCollector::visit(VariableDeclaration const& _node)
+{
+	if (&_node == &m_declaration)
+		m_resultingReferences.emplace_back(_node.nameLocation(), DocumentHighlightKind::Write);
+
+	return true;
+}
+
+bool ReferenceCollector::visitNode(ASTNode const& _node)
+{
+	if (&_node == &m_declaration)
+	{
+		if (auto const* declaration = dynamic_cast<Declaration const*>(&_node))
+			m_resultingReferences.emplace_back(declaration->nameLocation(), m_kind);
```

Comment:
Do I understand correctly that this counts a declaration as referencing itself? Do we really want that? This way you'll always find some references to a declaration even if it's in fact unused.

Might make sense for a variable declaration with an initializer but probably not for a declaration without one, contract definition or function definition...

Reply:
you want this, not just because the C++ LSP seems to highlight the declaration itself, too, but because you can later reuse the ReferenceCollector for symbolic renaming of function definition or contract definition and their uses, too.